src/main/scala/mesosphere/marathon/core/election/impl/CuratorElectionService.scala

```diff
@@ -67,13 +68,54 @@ class CuratorElectionService(
     maybeLatch.get.start()
   }
 
+  override protected def startLeadership(abdicate: Abdicator): Unit = {
+    maybeLatch match {
+      case None => super.startLeadership(abdicate)
+      case Some(l) =>
+        l.synchronized {
+          super.startLeadership(abdicate)
+        }
+    }
+  }
+
+  override protected def stopLeadership(): Unit = {
+    maybeLatch match {
+      case None => super.stopLeadership
+      case Some(l) =>
+        l.synchronized {
```

Comment:
The deadlocks we were seeing were between the latch and the MarathonSchedulerService.  So we synchronize on the latch at this spot before we synchronize on the the MarathonSchedulerService to prevent the deadlock.  The latch is synchronized from setLeadership in org.apache.curator/curator-recipes/srcs/curator-recipes-2.10.0-sources.jar!/org/apache/curator/framework/recipes/leader/LeaderLatch.java.  This was an easier fix than trying to change that LeaderLatch.java code.